cpp/include/cuspatial/experimental/detail/linestring_distance.cuh

```diff
@@ -0,0 +1,284 @@
+                                               vec_2d<T> const& a,
+                                               vec_2d<T> const& b)
+{
+  auto ab        = b - a;
+  auto ac        = c - a;
+  auto l_squared = dot(ab, ab);
+  if (l_squared == 0) { return dot(ac, ac); }
+  auto r  = dot(ac, ab);
+  auto bc = c - b;
+  // If the projection of `c` is outside of segment `ab`, compute point-point distance.
+  if (r <= 0 or r >= l_squared) { return std::min(dot(ac, ac), dot(bc, bc)); }
+  auto p  = a + (r / l_squared) * ab;
+  auto pc = c - p;
+  return dot(pc, pc);
+}
+
+/**
+ * @internal
+ * @brief Computes shortest distance between two segments (ab and cd) that
+ * doesn't intersect.
+ */
+template <typename T>
+T __device__ segment_distance_no_intersect_or_colinear(vec_2d<T> const& a,
+                                                       vec_2d<T> const& b,
+                                                       vec_2d<T> const& c,
+                                                       vec_2d<T> const& d)
+{
+  auto dist_sqr = std::min(std::min(point_to_segment_distance_squared(a, c, d),
+                                    point_to_segment_distance_squared(b, c, d)),
+                           std::min(point_to_segment_distance_squared(c, a, b),
+                                    point_to_segment_distance_squared(d, a, b)));
+  return dist_sqr;
+}
+
+/**
+ * @internal
+ * @brief Computes shortest distance between two segments.
+ *
+ * If two segments intersect, the distance is 0. Otherwise compute the shortest point
+ * to segment distance.
+ */
+template <typename T>
+T __device__ squared_segment_distance(vec_2d<T> const& a,
+                                      vec_2d<T> const& b,
+                                      vec_2d<T> const& c,
+                                      vec_2d<T> const& d)
+{
+  auto ab    = b - a;
+  auto cd    = d - c;
+  auto denom = det(ab, cd);
+
+  if (denom == 0) {
+    // Segments parallel or collinear
+    return segment_distance_no_intersect_or_colinear(a, b, c, d);
+  }
+
+  auto ac               = c - a;
+  auto r_numer          = det(ac, cd);
+  auto denom_reciprocal = 1 / denom;
+  auto r                = r_numer * denom_reciprocal;
+  auto s                = det(ac, ab) * denom_reciprocal;
+  if (r >= 0 and r <= 1 and s >= 0 and s <= 1) { return 0.0; }
+  return segment_distance_no_intersect_or_colinear(a, b, c, d);
+}
+
+/**
+ * @internal
+ * @brief The kernel to compute point to linestring distance
+ *
+ * Each thread of the kernel computes the distance between a segment in a linestring in pair 1
+ * to a linestring in pair 2. For a segment in pair 1, the linestring index is looked up from
+ * the offset array and mapped to the linestring in the pair 2. The segment is then computed
+ * with all segments in the corresponding linestringin pair 2. This forms a local minima of the
+ * shortest distance, which is then combined with other segment results via an atomic operation
```

Comment:
Atomics are really an implementation detail IMO.
```suggestion
 * shortest distance, which is then combined with other segment results via an operation
```

Reply:
The whole comment is implementation details. This is an internal function. Note the `@internal` at the top of the function comment. I think the word atomic should stay here, because algorithmically it doesn't make sense to just say "an operation".

Reply:
Fair enough, let's leave it then.